### DISCUSSION TITLE
Disambiguating attributes and braced style spans

### DISCUSSION THREAD

**Discussion by bpj** [jgm/djot#184] [Ideas]
The situation that bracketed styles like `{_emph_}` and attributes use the same kind of (curly) brackets is a bit problematic since it precludes using punctuation characters which have special meaning in attributes as style indicators and vice versa.

I’m thinking that maybe it could be possible to double the braces on style spans to disambiguate, i.e. `{{_emph_}}`, `{{#some style#}}`. I know that djot rightly frowns on doubled delimiters like `**foo**`, but IMHO braces are different because they are paired (the opening and closing delimiters are different chars) and thus it is possible to nest balanced braces to any depth, which is why the braced style span syntax exists in the first place, so I think this would be relatively benign. Naturally

1.  The doubled braces would only be *required* for those potential span markers which coincide with attribute syntax (`#`, `.`, `%`, `=`, `"`…) but they would be *allowed* with any style span.
2.  An opening delimiter with double brace could only be closed by a closing delimiter with double braces and ditto for delimiters with single braces (only `_}}` can close `{{_` and only `_}` can close `{_`), because that is **much** less confusing.

The parser(s) would have to keep track of single and double braces separately, but probably the two could share most code anyway. E.g. in Lua you could (for the sake of argument) use

``` lua
local start, stop, double, marker = str:find("^%{(%{?)([^%{%}%P])", pos)
```

and stash the info about the match in different places depending on whether `#double == 0` or not.

It would also be possible make some of the more esoteric span styles which currently require braces to require double braces, freeing up their marker characters for use in attributes, e.g. `[foo]{-lang}`/`[foo]{+role}` vs. `{{-del-}}`/`{{+ins+}}`.

I think this would be beneficial since the number of available ASCII punctuation characters is quite limited.

**Comment by bpj** [comment]
See also #174

**Comment by jgm** [comment]
This does add a lot of complexity to the mental model. People have to remember which characters need `{{` and which need `{`.  So I'm not currently in favor.

We could think about the nature of the problem.  `=` is used in attributes, but never at the beginning or end, so it's not a source of ambiguity and in fact we already use `{=` for highlighted text.  `#`, `.`, and `%` are definitely ruled out by their role in attributes.  We could perhaps reconsider `%`.  Currently this is an attribute comment, which can be used to put comments in a document.  But there are some drawbacks to this as a general comment syntax -- in particular, it doesn't allow blank lines in comments (#67).  If we had a distinct syntax for comments, then perhaps we could free up `%` to work with `{`.  Maybe something like https://github.com/jgm/djot/issues/67#issuecomment-1305033891.
